src/Avalonia.Controls/ColorPicker/ColorSpectrum/ColorSpectrum.cs

```diff
@@ -0,0 +1,1572 @@
+                _sizingGrid == null ||
+                _inputTarget == null ||
+                _spectrumRectangle == null ||
+                _spectrumEllipse == null ||
+                _spectrumOverlayRectangle == null ||
+                _spectrumOverlayEllipse == null
+                /*|| SharedHelpers.IsInDesignMode*/)
+            {
+                return;
+            }
+
+            // We want ColorSpectrum to always be a square, so we'll take the smaller of the dimensions
+            // and size the sizing grid to that.
+            double minDimension = Math.Min(_layoutRoot.Bounds.Width, _layoutRoot.Bounds.Height);
+
+            if (minDimension == 0)
+            {
+                return;
+            }
+
+            _sizingGrid.Width = minDimension;
+            _sizingGrid.Height = minDimension;
+
+            if (_sizingGrid.Clip is RectangleGeometry clip)
+            {
+                clip.Rect = new Rect(0, 0, minDimension, minDimension);
+            }
+
+            _inputTarget.Width = minDimension;
+            _inputTarget.Height = minDimension;
+            _spectrumRectangle.Width = minDimension;
+            _spectrumRectangle.Height = minDimension;
+            _spectrumEllipse.Width = minDimension;
+            _spectrumEllipse.Height = minDimension;
+            _spectrumOverlayRectangle.Width = minDimension;
+            _spectrumOverlayRectangle.Height = minDimension;
+            _spectrumOverlayEllipse.Width = minDimension;
+            _spectrumOverlayEllipse.Height = minDimension;
+
+            HsvColor hsvColor = HsvColor;
+            int minHue = MinHue;
+            int maxHue = MaxHue;
+            int minSaturation = MinSaturation;
+            int maxSaturation = MaxSaturation;
+            int minValue = MinValue;
+            int maxValue = MaxValue;
+            ColorSpectrumShape shape = Shape;
+            ColorSpectrumChannels channels = Channels;
+
+            // If min >= max, then by convention, min is the only number that a property can have.
+            if (minHue >= maxHue)
+            {
+                maxHue = minHue;
+            }
+
+            if (minSaturation >= maxSaturation)
+            {
+                maxSaturation = minSaturation;
+            }
+
+            if (minValue >= maxValue)
+            {
+                maxValue = minValue;
+            }
+
+            Hsv hsv = new Hsv(hsvColor);
+
+            // The middle 4 are only needed and used in the case of hue as the third dimension.
+            // Saturation and luminosity need only a min and max.
+            List<byte> bgraMinPixelData = new List<byte>();
+            List<byte> bgraMiddle1PixelData = new List<byte>();
+            List<byte> bgraMiddle2PixelData = new List<byte>();
+            List<byte> bgraMiddle3PixelData = new List<byte>();
+            List<byte> bgraMiddle4PixelData = new List<byte>();
+            List<byte> bgraMaxPixelData = new List<byte>();
```

Comment:
When I was implementing it last time with lists, I wasn't really thinking about ArrayPool usage. Or as we have in Avalonia PooledList (copied from another [library on github](https://github.com/jtmueller/Collections.Pooled)).

No need to reimplement it now, unless we will have serious perf problems with list solution.
If I recall this image isn't really recreated often with typical spectrul configuration.

Reply:
You are correct and the image isn't recalculated often. 

`Span<byte>` is something I considered but will probably just move to `byte[]` at some point. I'll comment elsewhere but I really don't like the double copy just to create a bitmap. In Uno I ended up writing my own ArrayList so I have access to the internal array directly which avoids one of those copies: https://github.com/unoplatform/uno/blob/master/src/Uno.UI/Microsoft/UI/Xaml/Controls/ColorPicker/ArrayList.cs

This is definitely something to keep in mind and perhaps revisit in the future. For now I didn't notice any major performance issues.